taxcalc/parameters_base.py

```diff
@@ -294,26 +293,75 @@ def _update(self, year_mods):
                          for i in range(0, num_years_to_expand)]
         else:
             inf_rates = None
-        paramvals = self._vals
+        used_names = set()  # set of used parameter names in year_mods dict
         for name, values in year_mods[year].items():
-            # determine inflation indexing status of parameter with name
+            # determine indexing status of parameter with name for year
             if name.endswith('_cpi'):
-                continue
-            if name in paramvals:
-                default_cpi = paramvals[name].get('cpi_inflated', False)
+                continue  # handle elsewhere in this method
+            if name in self._vals:
+                vals_indexed = self._vals[name].get('cpi_inflated', False)
             else:
-                default_cpi = False
-            cpi_inflated = year_mods[year].get(name + '_cpi', default_cpi)
+                msg = 'parameter name {} not in parameter values dictionary'
+                raise ValueError(msg.format(name))
+            name_plus_cpi = name + '_cpi'
+            if name_plus_cpi in year_mods[year].keys():
+                indexed = year_mods[year].get(name_plus_cpi)
+                self._vals[name]['cpi_inflated'] = indexed  # remember status
+                if name_plus_cpi in used_names:
+                    msg = 'parameter {} used twice in year_mods for year {}'
+                    raise ValueError(msg.format(name_plus_cpi, year))
+                else:
+                    used_names.add(name_plus_cpi)
+            else:
+                indexed = vals_indexed
             # set post-reform values of parameter with name
+            if name in used_names:
+                msg = 'parameter {} used twice in year_mods for year {}'
+                raise ValueError(msg.format(name, year))
+            else:
+                used_names.add(name)
             cval = getattr(self, name, None)
             if cval is None:
-                continue
+                msg = 'parameter {} in year_mods for year [] is unknown'
+                raise ValueError(msg.format(name, year))
             nval = self.expand_array(values,
-                                     inflate=cpi_inflated,
+                                     inflate=indexed,
                                      inflation_rates=inf_rates,
                                      num_years=num_years_to_expand)
-            cval[(self.current_year - self.start_year):] = nval
-        self.set_year(self._current_year)
+            cval[(year - self.start_year):] = nval
+        # handle unused parameter names, which all end in _cpi
+        for name, values in year_mods[year].items():
```

Comment:
@talumbau, You are right to say that we should avoid the second loop when possible and, if it is not possible to avoid it altogether, should loop through only the unused names.  I made some code changes that do this.  Let me know what you think.